tools/licenses/lib/formatter.dart

```diff
@@ -0,0 +1,164 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:core' hide RegExp;
+
+import 'patterns.dart';
+
+String reformat(String body) {
+  final List<String> lines = body.split('\n');
+  // The order of the following checks is important. If any test changes
+  // something that could affect an earlier test, then we should "continue" back
+  // to the top of the loop.
+  while (true) {
+    // Strip full-line decorations, e.g. horizontal lines, and trailing whitespace.
+    for (int index = 0; index < lines.length; index += 1) {
+      if (lines[index].endsWith(' ')) {
+        lines[index] = lines[index].trimRight();
+      }
+      if (fullDecorations.matchAsPrefix(lines[index]) != null) {
+        if (index == 0 || lines[index].length != lines[index - 1].length) {
+          // (we leave the decorations if it's just underlining the previous line)
+          lines[index] = '';
+        }
+      }
+    }
+    // Remove leading and trailing blank lines
+    while (lines.isNotEmpty && lines.first == '') {
+      lines.removeAt(0);
+    }
+    while (lines.isNotEmpty && lines.last == '') {
+      lines.removeLast();
+    }
+    if (lines.isEmpty) {
+      // We've stripped everything, give up.
+      return '';
+    }
+    // Try stripping leading indentation
+    String? prefix;
+    bool removeTrailingBlockEnd = false;
+    if (lines.first.startsWith('/*') && lines.last.startsWith(' *')) {
+      // in addition to the leadingDecorations, we also support one specific
+      // kind of multiline decoration, the /*...*/ block comment
+      prefix = ' *';
+      removeTrailingBlockEnd = true;
+    } else {
+      prefix = leadingDecorations.matchAsPrefix(lines.first)?.group(0);
+    }
+    if (prefix != null && lines.skip(1).every((String line) => line.startsWith(prefix!) || prefix.startsWith(line))) {
+      final int prefixLength = prefix.length;
+      for (int index = 0; index < lines.length; index += 1) {
+        final String line = lines[index];
+        if (line.length > prefixLength) {
+          lines[index] = line.substring(prefixLength);
+        } else {
+          lines[index] = '';
+        }
+      }
+      if (removeTrailingBlockEnd) {
+        if (lines.last == '/') {
+          // this removes the line with the trailing "*/" when we had a "/*" at the top
+          lines.removeLast();
+        }
+      }
+      continue; // go back to top since we may have more blank lines to strip now
+    }
+    // Try removing leading spaces
+    // (we know that this loop terminates before the end of the block because
+    // otherwise the previous section would have stripped a common prefix across
+    // the entire block)
```

Comment:
added:
```
    // For example, this will change: 
    // 
    // 	   foo 
    // 	 bar 
    // 
    // ...into: 
    // 
    // 	 foo 
    // 	 bar 
    // 
```